### PR TITLE
[PLA-4568] Convert training data to list of data sources

### PR DESCRIPTION
**Pull request by latture**
This PR converts training data from a single source to an optional list of data sources. A training data parameter has been added to the graph predictor. Any data defined by the graph will be shared by all sub-predictors. Training data for other predictors is optional, so training data does not need to be re-defined for all predictors in a graph.

This PR must be released after corresponding backend changes: [platform-backend #1329](https://github.com/CitrineInformatics/platform-backend/pull/1329).

### PR Type:
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/citrine/informatics/predictors.py [hunk 1]
```diff
@@ -83,8 +83,13 @@ class SimpleMLPredictor(Serializable['SimplePredictor'], Predictor):
         Descriptors that represent outputs of relations
     latent_variables: list[Descriptor]
         Descriptors that are predicted from inputs and used when predicting the outputs
-    training_data: DataSource
-        Source of the training data, which can be either a CSV or an Ara table
+    training_data: Optional[List[DataSource]]
+        Sources of training data. Each can be either a CSV or an Ara table.
+        Candidates from multiple data sources will be combined into a flattened list and deduplicated by uid
```
Comment (maxhutch) [inline src/citrine/informatics/predictors.py:88]: Should some or all of this description be included in the user guide as well?

Comment (latture) [reply src/citrine/informatics/predictors.py:88]: That's a good idea. I added a training data section to the user docs.

File: src/citrine/informatics/predictors.py [hunk 2]
```diff
@@ -66,6 +67,28 @@ def get_type(cls, data) -> Type['Predictor']:
...
+            A list of data sources
+        """
+        if training_data is None:
+            return []
+        if not isinstance(training_data, Iterable):
+            warn("Specifying training data as a single data source is deprecated. "
+                 "Please use a list of data sources to create {} instead.".format(self),
+                 DeprecationWarning)
+            return [training_data]
+        return training_data
```
Comment (latture) [inline src/citrine/informatics/predictors.py:90]: This is a new function to preserve backwards compatibility.

File: docs/source/workflows/predictors.rst [hunk 3]
```diff
@@ -481,13 +482,29 @@ This predictor will compute 2 responses, ``solute share in simple mixture`` and
...
 A :doc:`predictor report <predictor_reports>` describes a machine-learned model, for example its settings and what features are important to the model. 
 It does not include performance metrics. To learn more about performance metrics, please see :doc:`PerformanceWorkflows <performance_workflows>`.
 
-Data Sources
+Training data
 -------------
 
-A :doc:`data source <data_sources>` references the training data used by a predictor.
+Training data are defined by a list of :doc:`data sources <data_sources>`.
+When multiple data sources are specified, data from all sources is combined into a flattened list of candidates and deduplicated by uid and identifiers prior to training a predictor.
```
Comment (maxhutch) [inline docs/source/workflows/predictors.rst:495]:
```suggestion
When multiple data sources are specified, data from all sources is combined into a flattened list and deduplicated prior to training a predictor.
```

File: src/citrine/informatics/predictors.py [hunk 4]
```diff
@@ -66,6 +67,28 @@ def get_type(cls, data) -> Type['Predictor']:
...
+        training_data: Optional[Union[DataSource, List[DataSource]]]
+            Either a single data source, list of data sources or ``None``
+        Returns
+        -------
+        List[DataSource]
+            A list of data sources
+        """
+        if training_data is None:
+            return []
+        if not isinstance(training_data, Iterable):
```
Comment (maxhutch) [inline src/citrine/informatics/predictors.py:85]: `dict` is also Iterable.  Why not check `isinstance(training_data, DataSource)`?

File: src/citrine/informatics/predictors.py [hunk 5]
```diff
@@ -66,6 +67,28 @@ def get_type(cls, data) -> Type['Predictor']:
                 'Must be in {}.'.format(data['config']['type'], type_dict.keys())
             )
 
+    def _wrap_training_data(self, training_data: Optional[Union[DataSource, List[DataSource]]]) -> List[DataSource]:
```
Comment (maxhutch) [inline src/citrine/informatics/predictors.py:70]: Could this be a static method?

Comment (latture) [reply src/citrine/informatics/predictors.py:70]: I need access to `self` to issue a warning message that's specific to the instance.

File: src/citrine/informatics/predictors.py (hunk 4 above)
Comment (latture) [reply src/citrine/informatics/predictors.py:85]: That's a good idea. I'll update it.

**Pull request merged**